include/zephyr/usb/class/usbd_hid.h

```diff
@@ -0,0 +1,209 @@
+ *         | ------------------- |       |
+ *         |                     |       |
+ *         |   HID Device user   |       |
+ *         |      callbacks      |       |
+ *         |      handlers       |       |
+ *         +---------------------+       |
+ *                    ^                  |  HID Device Driver API:
+ *                    |                  |
+ *  set_protocol()    |                  |  register_app()
+ *  get_report()      |                  |  submit_report(
+ *  ....              |                  |  ...
+ *                    v                  |
+ *         +---------------------+       |
+ *         |                     |       |
+ *         |      HID Device     |       |
+ *         |     "bottom half"   |<------+
+ *         |    USB HID class    |
+ *         |    implementation   |
+ *         |                     |
+ *         |                     |
+ *         +---------------------+
+ *                    ^
+ *                    v
+ *          +--------------------+
+ *          |                    |
+ *          |     USB Device     |
+ *          |       Support      |
+ *          |                    |
+ *          +--------------------+
+ */
+
+/**
+ * @brief HID device user callbacks
+ *
+ * Each device depends on a user part that handles feature, input, and output
+ * report processing according to the device functionality described by the
+ * report descriptor. Which callbacks must be implemented depends on the device
+ * functionality. The USB device part of the HID device, cannot interpret
+ * device specific report descriptor and only handles USB specific parts,
+ * transfers and validation of requests, all reports are opaque to it.
+ * Callbacks are called from the USB device stack thread and must not block.
+ */
+struct hid_device_ops {
+	/**
+	 * The interface ready callback is called with the ready argument set
+	 * to true when the corresponding interface is part of the active
+	 * configuration and the application can e.g. begin submitting input
+	 * reports, and with the argument set to false when the interface is no
+	 * longer active. This callback is optional.
+	 */
+	void (*iface_ready)(const struct device *dev, const bool ready);
+
+	/**
+	 * Callback to get a feature, input, or output reports, specified by
+	 * type an id, from the HID device to the host. This callback is called
+	 * for the HID Get Report request. If there is no report ID in the
+	 * report descriptor, the id argument is zero. The callback
+	 * implementation must check the arquments, such as whether the report
+	 * type is supported, and return a nonzero value to indicate an
+	 * unsupported type or an error.
+	 */
+	int (*get_report)(const struct device *dev,
+			  const uint8_t type, const uint8_t id,
+			  const uint16_t len, uint8_t *const buf);
+
+	/**
+	 * Callback to set a feature, input, or output request, specified by
+	 * type an id, from the not to the HID device. This callback is called
```

Comment:
While the "type an id" seems to be simply missing "d", i.e. should be "type and id", I cannot quite grasp what "from the not to the HID device" is supposed to mean.